prometheus.lua

```diff
@@ -320,260 +140,374 @@ local function check_metric_and_label_names(metric_name, label_names)
+      key[i+2] = string.format("%sle=\"%s\"}", bucket_pref, self.bucket_format:format(buc))
+    end
+    -- Last bucket. Note, that the label value is "Inf" rather than "+Inf"
+    -- required by Prometheus. This is necessary for this bucket to be the last
+    -- one when all metrics are lexicographically sorted. "Inf" will get replaced
+    -- by "+Inf" in Prometheus:collect().
+    key[self.bucket_count+3] = string.format("%sle=\"Inf\"}", bucket_pref)
+  else
+    key = full_metric_name(self.name, self.label_names, label_values)
+  end
+  t[KEY_METRIC] = key
+  return key
 end
 
--- Register a counter.
---
--- Args:
---   name: (string) name of the metric. Required.
---   description: (string) description of the metric. Will be used for the HELP
---     comment on the metrics page. Optional.
---   label_names: array of strings, defining a list of metrics. Optional.
---
--- Returns:
---   a Counter object.
-function Prometheus:counter(name, description, label_names)
-  if not self.initialized then
-    ngx.log(ngx.ERR, "Prometheus module has not been initialized")
-    return
-  end
+local ERR_MSG_COUNTER_NOT_INITIALIZED = "counter not initialied"
 
-  local err = check_metric_and_label_names(name, label_names)
-  if err ~= nil then
-    self:log_error(err)
+local function inc(self, value, label_values)
+  -- counter is not allowed to decrease
+  if self.typ == TYPE_COUNTER and value and value < 0 then
+    self._log_error_kv(self.name, value, "Value should not be negative")
     return
   end
 
-  if self.registered[name] then
-    self:log_error("Duplicate metric " .. name)
+  local k, _, err
+  k, err = lookup_or_create(self, label_values)
+  if err then
+    self._log_error(err)
     return
   end
-  self.registered[name] = true
-  self.help[name] = description
-  self.type[name] = "counter"
-
-  return Counter:new{name=name, label_names=label_names, prometheus=self}
+  -- FIXME: counter is initialized in init_worker while metrics are initiliazed
+  -- in init phase
+  local c = self._counter
+  if not c then
+    c = self.parent._counter
+    if not c then
+      self._log_error(ERR_MSG_COUNTER_NOT_INITIALIZED)
+      return
+    end
+    self._counter = c
+  end
+  _, err = c:incr(k, value, 0)
+  if err then
+    self._log_error(err)
+  end
 end
 
--- Register a gauge.
---
--- Args:
---   name: (string) name of the metric. Required.
---   description: (string) description of the metric. Will be used for the HELP
---     comment on the metrics page. Optional.
---   label_names: array of strings, defining a list of metrics. Optional.
---
--- Returns:
---   a Gauge object.
-function Prometheus:gauge(name, description, label_names)
-  if not self.initialized then
-    ngx.log(ngx.ERR, "Prometheus module has not been initialized")
+local function del(self, label_values)
+  local k, _, err
+  k, err = lookup_or_create(self, label_values)
+  if err then
+    self._log_error(err)
     return
   end
+  _, err = self._dict:delete(k)
```

Comment:
This is going to race with counter flushes. Should we wait here for `sync_interval` like we do in `reset()`?